src/boot/efi/random-seed.c

```diff
@@ -271,51 +212,86 @@ EFI_STATUS process_random_seed(EFI_FILE *root_dir, RandomSeedMode mode) {
                 return log_error_status_stall(EFI_INVALID_PARAMETER, L"Random seed file is too large.");
 
         seed = xmalloc(size);
-
         rsize = size;
         err = handle->Read(handle, &rsize, seed);
         if (err != EFI_SUCCESS)
                 return log_error_status_stall(err, L"Failed to read random seed file: %r", err);
-        if (rsize != size)
+        if (rsize != size) {
+                explicit_bzero_safe(seed, rsize);
                 return log_error_status_stall(EFI_PROTOCOL_ERROR, L"Short read on random seed file.");
+        }
+
+        sha256_process_bytes(&size, sizeof(size), &hash);
+        sha256_process_bytes(seed, size, &hash);
+        explicit_bzero_safe(seed, size);
 
         err = handle->SetPosition(handle, 0);
         if (err != EFI_SUCCESS)
                 return log_error_status_stall(err, L"Failed to seek to beginning of random seed file: %r", err);
 
-        /* Request some random data from the UEFI RNG. We don't need this to work safely, but it's a good
-         * idea to use it because it helps us for cases where users mistakenly include a random seed in
-         * golden master images that are replicated many times. */
-        (void) acquire_rng(size, &rng); /* It's fine if this fails */
-
         /* Let's also include the UEFI monotonic counter (which is supposedly increasing on every single
          * boot) in the hash, so that even if the changes to the ESP for some reason should not be
          * persistent, the random seed we generate will still be different on every single boot. */
         err = BS->GetNextMonotonicCount(&uefi_monotonic_counter);
         if (err != EFI_SUCCESS)
                 return log_error_status_stall(err, L"Failed to acquire UEFI monotonic counter: %r", err);
+        sha256_process_bytes(&uefi_monotonic_counter, sizeof(uefi_monotonic_counter), &hash);
 
-        /* Calculate new random seed for the disk and what to pass to the kernel */
-        err = mangle_random_seed(seed, rng, size, system_token, system_token_size, uefi_monotonic_counter, &new_seed, &for_kernel);
-        if (err != EFI_SUCCESS)
-                return err;
+        /* hash_key = HASH(hash) */
+        sha256_finish_ctx(&hash, hash_key);
+
+        /* hash = hash_key || 0 */
+        sha256_init_ctx(&hash);
+        sha256_process_bytes(hash_key, sizeof(hash_key), &hash);
+        usage_idx = 0;
+        sha256_process_bytes(&usage_idx, sizeof(usage_idx), &hash);
+        /* random_bytes = HASH(hash) */
+        sha256_finish_ctx(&hash, random_bytes);
 
         /* Update the random seed on disk before we use it */
+        size = sizeof(random_bytes);
+        if (size < info->FileSize) {
+                info->FileSize = size;
+                err = handle->SetInfo(handle, &GenericFileInfo, info->Size, info);
```

Comment:
Wouldn't this likely make the above seed file size checks fail? Originally, this code was designed defensively so that no fs changes (beside overriding already used blocks) were needed.

Reply:
No, I think you might have misread this part. It simply truncates the file in case it was really big before.

Reply:
If it was too big it would've been rejected anyway?

Reply:
To put it differently: The code requires the seed file to be within `RANDOM_MAX_SIZE_MIN` and `RANDOM_MAX_SIZE_MAX`. If that upper bound makes no sense anymore, it should really be changed and userspace should be responsible for fixing the file size…

Reply:
No, the upper bound is 1024, so it won't be rejected earlier.

Adjusting the upper bound doesn't make sense; that'll just invalidate perfectly acceptable seed files that tools might be creating or might already exist. Hashing in larger files is perfectly valid here. It's just that there's no point in us creating new ones larger than 32 bytes, so we don't.